server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -1817,7 +1822,147 @@ public void publishVolumeCreationUsageEvent(Volume volume) {
+                    } else if (jobResult instanceof ResourceAllocationException) {
+                        throw (ResourceAllocationException)jobResult;
+                    } else if (jobResult instanceof Throwable) {
+                        throw new RuntimeException("Unexpected exception", (Throwable)jobResult);
+                    }
+                }
+
+                // retrieve the entity url from job result
+                if (jobResult != null && jobResult instanceof Pair) {
+                    return (Pair<String, String>) jobResult;
+                }
+
+                return null;
+            }
+        } else {
+            CheckAndRepairVolumePayload payload = new CheckAndRepairVolumePayload(repair);
+            VolumeInfo volumeInfo = volFactory.getVolume(volumeId);
+            volumeInfo.addPayload(payload);
+
+            Pair<String, String> result = volService.checkAndRepairVolume(volumeInfo);
+            return result;
+        }
+    }
+
+    protected void validationsForCheckVolumeOperation(VolumeVO volume) {
+        Account caller = CallContext.current().getCallingAccount();
+        _accountMgr.checkAccess(caller, null, true, volume);
+
+        Long volumeId = volume.getId();
+        Long vmId = volume.getInstanceId();
+        if (vmId != null) {
+            UserVmVO vm = _userVmDao.findById(vmId);
+            if (vm == null) {
+                throw new InvalidParameterValueException(String.format("VM not found, please check the VM to which this volume %d is attached", volumeId));
+            }
+
+            _accountMgr.checkAccess(caller, null, true, vm);
+
+            if (vm.getState() != State.Stopped) {
+                throw new InvalidParameterValueException(String.format("VM to which the volume %d is attached should be in stopped state", volumeId));
+            }
+        }
+
+        if (volume.getState() != Volume.State.Ready) {
+            throw new InvalidParameterValueException(String.format("VolumeId: %d is not in Ready state", volumeId));
+        }
+
+        HypervisorType hypervisorType = _volsDao.getHypervisorType(volume.getId());
+        if (!HypervisorType.KVM.equals(hypervisorType)) {
+            throw new InvalidParameterValueException(String.format("Check and Repair volumes is supported only for KVM hypervisor"));
+        }
+    }
+
+    private Pair<String, String> orchestrateCheckAndRepairVolume(Long volumeId, String repair) {
+
+        VolumeInfo volume = volFactory.getVolume(volumeId);
+
+        if (volume == null) {
+            throw new InvalidParameterValueException("Checking volume and repairing failed due to volume:" + volumeId + " doesn't exist");
+        }
+
+        if (volume.getState() != Volume.State.Ready) {
+            throw new InvalidParameterValueException("VolumeId: " + volumeId + " is not in " + Volume.State.Ready + " state but " + volume.getState() + ". Cannot check and repair the volume.");
+        }
```

Comment:
This is already checked at line 1907 in the validationsForCheckVolumeOperation method